plugins/de.cau.cs.kieler.klighd/src/de/cau/cs/kieler/klighd/filtering/parser/SemanticFilterRuleParser.java

```diff
@@ -0,0 +1,296 @@
+
+import static java.util.Map.entry;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+import java.util.StringTokenizer;
+
+import de.cau.cs.kieler.klighd.filtering.AndConnective;
+import de.cau.cs.kieler.klighd.filtering.FalseConnective;
+import de.cau.cs.kieler.klighd.filtering.GreaterEqualsConnective;
+import de.cau.cs.kieler.klighd.filtering.GreaterThanConnective;
+import de.cau.cs.kieler.klighd.filtering.LessEqualsConnective;
+import de.cau.cs.kieler.klighd.filtering.LessThanConnective;
+import de.cau.cs.kieler.klighd.filtering.LogicEqualConnective;
+import de.cau.cs.kieler.klighd.filtering.NegationConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericAdditionConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericConstantConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericDivisionConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericEqualConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericMultiplicationConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericNotEqualConnective;
+import de.cau.cs.kieler.klighd.filtering.NumericResult;
+import de.cau.cs.kieler.klighd.filtering.NumericSubtractionConnective;
+import de.cau.cs.kieler.klighd.filtering.OrConnective;
+import de.cau.cs.kieler.klighd.filtering.SemanticFilterRule;
+import de.cau.cs.kieler.klighd.filtering.SemanticFilterTag;
+import de.cau.cs.kieler.klighd.filtering.TrueConnective;
+
+/**
+ * @author mka
+ *
+ */
+public class SemanticFilterRuleParser {
+    
+    private abstract class Node {
+        public String token;
+        // current return type of AST
+        // -1: unknown
+        // 0: boolean
+        // 1: numeric
+        public int type;
+        public Node(String token, int type) {
+            this.token = token;
+            this.type = type;
+        }
+    }
+    
+    private class OperatorNode extends Node {
+        private List<Node> children;
+        
+        public List<Node> getChildren() { return children; }
+        
+        public OperatorNode(String token, int type) {
+            super(token, type);
+            this.children = new ArrayList<>();
+        }
+        public OperatorNode(String token, int type, List<Node> children) {
+            super(token, type);
+            this.children = children;
+        }
+        
+        public void addChild(Node child) {
+            this.children.add(child);
+        }
+    }
+    
+    private class OperandNode extends Node {
+
+        public OperandNode(String token, int type) {
+            super(token, type);
+        }
+    }
+    
+    public class InvalidSyntaxException extends Exception {
+
+        public InvalidSyntaxException(String message) {
+            super(message);
+        }
+    }
+    
+    private SemanticFilterRule convertAST(Node ast) throws InvalidSyntaxException {
+        try {
+            SemanticFilterRule result;
+            switch(ast.token) {
```

Comment:
Is this switch exhaustive? I remember the if/then connectives, should they be included in this language as well or not?

Reply:
This doesn't include those connectives at the moment because the parsing is a little more complicated. In principle they don't extend the expressiveness of the language but of course it would be nicer to have a one-to-one correspondence of the features